includes/admin/extensions/class-extensions-api.php

```diff
@@ -0,0 +1,201 @@
+
+		// If the data is "fresh" and what we want exists, return it.
+		if ( ! $is_stale ) {
+			if ( $item_id && ! empty( $option[ $item_id ] ) ) {
+				return $product_data->fromArray( $option[ $item_id ] );
+			} elseif ( ! empty( $option['timeout'] ) ) {
+				unset( $option['timeout'] );
+
+				return $option;
+			}
+		}
+
+		// Get all of the product data.
+		$all_product_data = $this->get_all_product_data();
+
+		// If no product data was retrieved, let the option sit for an hour.
+		if ( ! $all_product_data ) {
+			$data = array(
+				'timeout' => strtotime( '+1 hour', time() ),
+			);
+			if ( $option && $is_stale ) {
+				$data = array_merge( $option, $data );
+			}
+			update_option(
+				$option_name,
+				$data,
+				false
+			);
+
+			if ( $item_id && ! empty( $option[ $item_id ] ) ) {
+				return $product_data->fromArray( $option[ $item_id ] );
+			}
+			unset( $option['timeout'] );
+
+			return $option;
+		}
+
+		$value = array(
+			'timeout' => strtotime( '+1 week', time() ),
+		);
+		if ( $item_id && ! empty( $all_product_data->$item_id ) ) {
+			$item              = $all_product_data->$item_id;
+			$value[ $item_id ] = $this->get_item_data( $item );
+		} elseif ( in_array( $key, array( 'category', 'tag' ), true ) ) {
+			$term_id = $body[ $key ];
+			foreach ( $all_product_data as $item_id => $item ) {
+				if ( 'category' === $key && ( empty( $item->categories ) || ! in_array( $term_id, $item->categories, true ) ) ) {
+					continue;
+				} elseif ( 'tag' === $key && ( empty( $item->tags ) || ! in_array( $term_id, $item->tags, true ) ) ) {
+					continue;
+				}
+				$value[ $item_id ] = $this->get_item_data( $item );
+			}
+		}
+
+		update_option( $option_name, $value, false );
+		unset( $value['timeout'] );
+
+		return $item_id && ! empty( $value[ $item_id ] ) ? $product_data->fromArray( $value[ $item_id ] ) : $value;
+	}
+
+	/**
+	 * Gets all of the product data, either from a tranient or an API request.
+	 *
+	 * @since 2.11.x
+	 * @return object|false
+	 */
+	private function get_all_product_data() {
+		// Possibly all product data is in a transient. If it is, return it.
+		$all_product_data = get_transient( 'edd_all_extension_data' );
+		if ( $all_product_data ) {
+			return $all_product_data;
+		}
+
+		// Otherwise, query the API.
+		$url     = add_query_arg(
+			array(
+				'edd_action' => 'extension_data',
+			),
+			$this->get_products_url()
+		);
+		$request = wp_remote_get(
+			$url,
+			array(
+				'timeout'   => 15,
+				'sslverify' => true,
+			)
+		);
+
+		// If there was an API error, set timeout for 1 hour and return false.
+		if ( is_wp_error( $request ) || ( 200 !== wp_remote_retrieve_response_code( $request ) ) ) {
+			return false;
+		}
+
+		// Fresh data has been retrieved, so populate with fresh data.
+		$all_product_data = json_decode( wp_remote_retrieve_body( $request ) );
+
+		// Store all of the product data for one hour.
+		set_transient( 'edd_all_extension_data', $all_product_data, HOUR_IN_SECONDS );
```

Comment:
Wondering if this should be longer than an hour. :thinking:

Reply:
Chris asked me yesterday what was left for this PR and what I needed decisions from him on. I listed all of the "timings" that I've set up and asked him to review and ensure that each is a reasonable time in his mind. Specifically, I told him:

>
> * there is a transient on the `extension_data` request that expires after an hour. initially that was because the settings checks weren't working quite right and I was having four requests all at once. this way if someone goes to the marketing tab, the request for email marketing will also work for Reviews, for example
> * the extension data is saved for one week
> * if an extension data request fails but there is existing data, it's not deleted, but the expiration time is reset for an hour, so that stale data is used instead of having no data
> * if there is no extension data and the request fails then there is a timeout for an hour
> That first one is a transient but the specific extension data is saved as an option.

So I'm totally fine to change this to be longer but maybe we should see what he thinks. He is looking at caching the extension data request on the marketing site side as well.

Reply:
Okay I'm fine with whatever number he's happy with.

Reply:
My only concern here is the transient, if transients are turned off, we run the risk of sending a lot of traffic our way, even with it restricted to specific pages, it is easy for that to get forgotten in future releases. I'm going to try and setup a caching layer into this so it never touches the server, but even that will expire at some point and require hitting the web backend.

Reply:
Okay, I have updated the API class to not use any transients at all. I also adjusted the timings:
- `edd_all_extension_data` has a timeout of 1 hour if the API request fails; 4 hours if it succeeds
- Individual product options have a timeout of 1 hour if the API request fails; 1 week if it succeeds

If the individual product options have "expired", the stale data is merged with an updated 1 hour timeout. I did not do this for the all extension data, just set the timeout.

Reply:
I like that a lot better, should protect us from rouge checks.